File: opsmate/providers/openai.py
from __future__ import annotations
import json
import openai
from .base import BaseProvider, Response, ToolCall


class OpenAIProvider(BaseProvider):
    def __init__(self, api_key: str, model: str) -> None:
        self._client = openai.OpenAI(api_key=api_key)
        self._model = model

    @property
    def label(self) -> str:
        return f"openai/{self._model}"

    def chat(self, messages: list[dict], tool_schemas: list[dict], system: str = "") -> Response:
        msgs = messages.copy()
        if system:
            msgs = [{"role": "system", "content": system}] + msgs

        kwargs: dict = dict(model=self._model, messages=msgs)
        if tool_schemas:
            kwargs["tools"] = tool_schemas

        resp = self._client.chat.completions.create(**kwargs)
        msg = resp.choices[0].message

        calls: list[ToolCall] = []
        if msg.tool_calls:
            for tc in msg.tool_calls:
                calls.append(ToolCall(
                    id=tc.id,
                    name=tc.function.name,
                    arguments=json.loads(tc.function.arguments),
                ))

        return Response(
            content=msg.content or "",
            tool_calls=calls,
            input_tokens=resp.usage.prompt_tokens,
            output_tokens=resp.usage.completion_tokens,
            _raw=msg,
        )

    def append_response(self, messages: list[dict], response: Response) -> None:
        msg = response._raw
        entry: dict = {"role": "assistant", "content": msg.content or ""}
        if msg.tool_calls:
            entry["tool_calls"] = [
                {
                    "id": tc.id,
                    "type": "function",
                    "function": {"name": tc.function.name, "arguments": tc.function.arguments},
                }
                for tc in msg.tool_calls
            ]
        messages.append(entry)

    def append_tool_results(self, messages: list[dict], results: list[tuple[ToolCall, str]]) -> None:
        for call, result in results:
            messages.append({"role": "tool", "tool_call_id": call.id, "content": result})

    def build_tool_schemas(self, tools: list) -> list[dict]:
        return [
            {
                "type": "function",
                "function": {
                    "name": t.name,
                    "description": t.description,
                    "parameters": t.parameters,
                },
            }
            for t in tools
        ]

File: pkg/llm/openai.go
package llm

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/sashabaranov/go-openai"
)

// OpenAIClient implements Client using go-openai (also works with Ollama's
// OpenAI-compatible endpoint by setting a custom BaseURL).
type OpenAIClient struct {
	client *openai.Client
	model  string
}

// NewOpenAIClient creates a client using the standard OpenAI endpoint.
func NewOpenAIClient(apiKey, model string) *OpenAIClient {
	return &OpenAIClient{client: openai.NewClient(apiKey), model: model}
}

// NewOllamaClient creates a client pointing at a local Ollama instance.
func NewOllamaClient(baseURL, model string) *OpenAIClient {
	cfg := openai.DefaultConfig("ollama")
	cfg.BaseURL = baseURL + "/v1"
	return &OpenAIClient{client: openai.NewClientWithConfig(cfg), model: model}
}

func (o *OpenAIClient) ProviderName() string { return "openai" }

func (o *OpenAIClient) Complete(ctx context.Context, req Request) (Response, error) {
	msgs := make([]openai.ChatCompletionMessage, 0, len(req.Messages))
	if req.System != "" {
		msgs = append(msgs, openai.ChatCompletionMessage{
			Role:    openai.ChatMessageRoleSystem,
			Content: req.System,
		})
	}
	for _, m := range req.Messages {
		role := openai.ChatMessageRoleUser
		switch m.Role {
		case RoleAssistant:
			role = openai.ChatMessageRoleAssistant
		case RoleTool:
			role = openai.ChatMessageRoleTool
		}
		msgs = append(msgs, openai.ChatCompletionMessage{
			Role:       role,
			Content:    m.Content,
			ToolCallID: m.ToolCallID,
		})
	}

	creq := openai.ChatCompletionRequest{
		Model:       o.model,
		Messages:    msgs,
		Temperature: float32(req.Temperature),
	}
	if req.MaxTokens > 0 {
		creq.MaxTokens = req.MaxTokens
	}

	if len(req.Tools) > 0 {
		tools := make([]openai.Tool, len(req.Tools))
		for i, t := range req.Tools {
			schemaBytes, err := json.Marshal(t.InputSchema)
			if err != nil {
				return Response{}, fmt.Errorf("openai: marshal tool schema: %w", err)
			}
			tools[i] = openai.Tool{
				Type: openai.ToolTypeFunction,
				Function: &openai.FunctionDefinition{
					Name:        t.Name,
					Description: t.Description,
					Parameters:  json.RawMessage(schemaBytes),
				},
			}
		}
		creq.Tools = tools
		switch req.ToolChoice {
		case ToolChoiceRequired:
			creq.ToolChoice = "required"
		case ToolChoiceNone:
			creq.ToolChoice = "none"
		}
	}

	resp, err := o.client.CreateChatCompletion(ctx, creq)
	if err != nil {
		return Response{}, fmt.Errorf("openai: api error: %w", err)
	}

	if len(resp.Choices) == 0 {
		return Response{}, fmt.Errorf("openai: empty choices")
	}

	msg := resp.Choices[0].Message
	out := Response{
		Content: msg.Content,
		Model:   o.model,
		Usage: Usage{
			InputTokens:  resp.Usage.PromptTokens,
			OutputTokens: resp.Usage.CompletionTokens,
		},
	}
	for _, tc := range msg.ToolCalls {
		var input map[string]interface{}
		_ = json.Unmarshal([]byte(tc.Function.Arguments), &input)
		out.ToolCalls = append(out.ToolCalls, ToolCall{
			ID:    tc.ID,
			Name:  tc.Function.Name,
			Input: input,
		})
	}
	return out, nil
}
